c2corg_api/tests/views/test_document_delete.py

```diff
@@ -0,0 +1,562 @@
+        )
+        self.session.add(self.topo_map1)
+        self.session.flush()
+        self.session.add(TopoMapAssociation(
+            document=self.waypoint2, topo_map=self.topo_map1))
+        self.session.add(TopoMapAssociation(
+            document=self.waypoint3, topo_map=self.topo_map1))
+        self.session.add(TopoMapAssociation(
+            document=self.route2, topo_map=self.topo_map1))
+        self.session.add(TopoMapAssociation(
+            document=self.route3, topo_map=self.topo_map1))
+        self.session.flush()
+
+        self.area1 = Area(
+            area_type='range',
+            geometry=DocumentGeometry(
+                geom_detail='SRID=3857;POLYGON((611774.917032556 5706934.10657514,611774.917032556 5744215.5846397,642834.402570357 5744215.5846397,642834.402570357 5706934.10657514,611774.917032556 5706934.10657514))'  # noqa
+            )
+        )
+        self.session.add(self.area1)
+        self.session.flush()
+        self.session.add(AreaAssociation(
+            document=self.waypoint2, area=self.area1))
+        self.session.add(AreaAssociation(
+            document=self.waypoint3, area=self.area1))
+        self.session.add(AreaAssociation(
+            document=self.route2, area=self.area1))
+        self.session.add(AreaAssociation(
+            document=self.route3, area=self.area1))
+        self.session.flush()
+
+    def _add_association(self, parent_document, child_document):
+        association = Association.create(
+            parent_document=parent_document,
+            child_document=child_document)
+        self.session.add(association)
+        self.session.add(association.get_log(
+            self.global_userids['contributor']))
+
+    def _delete(self, document_id, expected_status):
+        headers = self.add_authorization_header(username='moderator')
+        return self.app.delete_json(
+            self._prefix + str(document_id),
+            headers=headers, status=expected_status)
+
+    def test_non_unauthorized(self):
+        self.app.delete_json(
+            self._prefix + str(self.waypoint1.document_id), {}, status=403)
+
+        headers = self.add_authorization_header(username='contributor')
+        return self.app.delete_json(
+            self._prefix + str(self.waypoint1.document_id), {},
+            headers=headers, status=403)
+
+    def test_non_existing_document(self):
+        self._delete(-9999999, 400)
+
+    def test_delete_main_waypoint(self):
+        """ Test that a main waypoint cannot be deleted.
+        """
+        response = self._delete(self.waypoint1.document_id, 400)
+        errors = response.json.get('errors')
+        self.assertEqual(len(errors), 1)
+        self.assertEqual(errors[0].get('name'), 'Bad Request')
```

Comment:
There is a convenience method to make testing for errors simpler. For example you could write this test as:

```python
self.assertErrorsContain(response.json, 'Bad Request', 'This waypoint cannot be deleted because it is a main waypoint.')
```